### ISSUE TITLE
Substitution list just shows 'Loading...'

### ISSUE THREAD

**Issue by Merseybuoy** [Woundorf/foxreplace#223]
Version 1.1.3 has been working fine for a few days. Now, when I open the options page, the page loads but all I get is an empty window where my lists should be, and a 'Loading...' message in the middle. Auto-replace isn't working but manual replacement is. 
Tried uninstalling (including settings), reinstalling, installing version 1.0 (which works), deleting storage.js and reinstalling. 
Using Firefox 56.0b12. Tried dropping back to b10, no change.
I can make the lists visible by importing list.json but it disappears again if I refresh the page.

EDIT: Just found that turning off multiprocess caused this. I'd turned it off because it broke another extension.

**Comment by Woundorf** [comment]
OK, thank you for the report and adding the cause later. I don't know why it doesn't work without multiprocess, but I'll try to find it if I have time.

**Comment by sqgl2** [comment]
I get the same bug but don't know what Merseybuoy means by turning off multiprocess. Does he mean he set affinity of FireFox to a single core? Anyhow, I did no such thing.

Where is the most recent json file stored so that I can load it manually? Or am I going to have to rely on an old json manual backup?

At first I noticed that FR was not working even though FR options were open in the tab. Manual sub failed too. Restarted FF and then options are in permanent "loading" state.

FF 57.0.1 (64bit), Win7, FR 2.1.0

**Comment by Merseybuoy** [comment]
@sqgl2. Firefox multiprocess can be enabled or disabled from About:Config. Loads of info on Google. However, I've long since moved on to version 2.1 of FoxReplace and Firefox 58beta and it's all working fine (with multiprocess enabled). As far as I can remember, I haven't had this 'loading' problem with the latest FoxReplace.

**Comment by sqgl2** [comment]
Upon comparison with an old saved FR json file I notice the above JS file is just one long line but is very similar except for additional header and tail stuff in the JS file. I then used Notepad++ with a plugin called NPPJSONViewer.dll and removed the superfluous header and tail of the JS file. Renamed it with a JSON extension and it worked.

Can we please have an easy way to do this within FR in future?

**Comment by sqgl2** [comment]
Woundorf I can give you the file which wouldn't load in FR if you like. I shouldn't have been allowed to save it in FR if it wasn't able to load it later. Surely this is a bug.

**Comment by Woundorf** [comment]
@sqgl2 That file you mention, storage.js, is created by Firefox (not FoxReplace) to store the extension's internal data, which includes the list and all the other settings. You can find equivalent files for all your installed extensions. Thus, it's not a bug nor a feature, it's just Mozilla's implementation.

**Comment by sqgl2** [comment]
Nevertheless the question remains: why am I allowed to save a file which is so corrupt it breaks FR upon a FF restart? Couldn't the check be done during the save and not just during the load?

**Comment by Woundorf** [comment]
I have tried importing a broken JSON file and even other file types and it gives an error and doesn't modify the list. Thus I don't know how you could save it.

**Comment by sqgl2** [comment]
Yes I got errors too upon importing it (after cropping FF's internal data JS file). But I am talking about a "load" not an "import", ie what happens when FF is restarted.

**Comment by Woundorf** [comment]
Yes, but if I understood correctly you mean load after saving corrupt data, and I suppose you mean saving it using FoxReplace's interface, not modifying the storage.js file on your own. And to save data the only way is with the Save button in the options page and that saves the data that is currently entered in that page. Thus you have had to enter the corrupt data somehow, and the only possibilities are manually creating substitutions or importing them from a file or url. From the provided editor it should be impossible to create corrupt data (if there is a way then it's a very serious bug), and when importing there are checks.

**Comment by sqgl2** [comment]
You understand correctly. I manually created bad substitutions by not only leaving out a closing parenthetical in a regex (which FR allows us to save) but by having the following entered accidentally by an automated script ([AutoHotKey](https://autohotkey.com/)) I had running in the background. I didn't notice and hit SAVE.

```
send {sleep 11}
WinActivate, sqgl@netpd
send {sleep 11}
send date{enter}
send {sleep 11}
send transcoder_64 status{enter}
send {sleep 11}
send transcoder_64 restart{enter}
send {sleep 11}
run D:\AUDIO\Foobar_64\foobar2000.exe /play
```
> From the provided editor it should be impossible to create corrupt data (if there is a way then it's a very serious bug)

Well it happened but would be difficult to replicate. So difficult it might be worth ignoring it for now, thanks for your response though. If it happens to anybody else they at least now have my tips for resurrecting and editing their last save.

**Comment by Woundorf** [comment]
Well, the unclosed parenthesis is really problematic because you can create it in the editor. I see you already reported it in #236 but I didn't know that it could cause this problem. I will fix it as soon as possible.

**Comment by Woundorf** [comment]
This should be fixed with e26922c. I'll close it for now, but if anyone can still reproduce it (after the next release) I'll reopen it.
